atd-etl/cris_import/create_fact_table_public_views.py

```diff
@@ -0,0 +1,584 @@
+            "opitem": "SFTP Endpoint Key",
+            "opfield": ".private key",
+            "opvault": VAULT_ID,
+            },
+        "bastion_ssh_private_key": {
+            "opitem": "RDS Bastion Key",
+            "opfield": ".private key",
+            "opvault": VAULT_ID,
+            },
+    }
+
+    # instantiate a 1Password client
+    client: Client = new_client(ONEPASSWORD_CONNECT_HOST, ONEPASSWORD_CONNECT_TOKEN)
+    # get the requested secrets from 1Password
+    return onepasswordconnectsdk.load_dict(client, REQUIRED_SECRETS)
+
+
+def make_crashes_view():
+    pg = get_pg_connection()
+    db = pg.cursor(cursor_factory=psycopg2.extras.DictCursor)
+    
+    db.execute("drop view if exists public.atd_txdot_crashes;")
+    pg.commit()
+
+    # ldm.atd_txdot_crashes columns we want to compute on the fly
+    # * latitude_primary
+    # * longitude_primary
+    
+    sql = """
+    with vz as (
+        select
+            'vz_facts' as schema, ordinal_position,
+            column_name, data_type, udt_name, character_maximum_length, numeric_precision,
+            numeric_precision, numeric_scale, is_generated, generation_expression, is_updatable
+        FROM information_schema.columns
+        WHERE true
+            AND table_schema = 'vz_facts'
+            AND table_name = 'atd_txdot_crashes'
+        order by ordinal_position
+        ), cris as (
+        SELECT
+            'cris' as schema, ordinal_position,
+            column_name, data_type, udt_name, character_maximum_length, numeric_precision,
+            numeric_precision, numeric_scale, is_generated, generation_expression, is_updatable
+        FROM information_schema.columns
+        WHERE true
+            AND table_schema = 'cris_facts'
+            AND table_name = 'atd_txdot_crashes'
+        order by ordinal_position
+        )
+    select vz.column_name as vz_column_name, cris.column_name as cris_column_name,
+        case when vz.column_name is not null then vz.schema
+            when cris.column_name is not null then cris.schema
+        end as schema,
+        case when vz.column_name is not null then vz.column_name
+            when cris.column_name is not null then cris.column_name
+        end as column_name
+    from vz
+        full outer join cris on (vz.column_name = cris.column_name)
+    where true 
+        and (vz.column_name not in ('crash_id') or cris.column_name not in ('crash_id'))
+    order by coalesce(vz.ordinal_position, cris.ordinal_position)
+    """
+    db.execute(sql)
+
+
+    view = """
+    create view public.atd_txdot_crashes as
+        select
+            cris_facts.atd_txdot_crashes.crash_id as crash_id,
+        """
+    columns = []
+    for column in db:
+        print("Column: ", column["vz_column_name"])
+        if column["vz_column_name"] == "longitude_primary":
+            columns.append("ST_X(COALESCE(vz_facts.atd_txdot_crashes.position, cris_facts.atd_txdot_crashes.position)) as longitude_primary")
```

Comment:
Everything you've mentioned here is right on the money. I hope we use the dynamic calculation of derived information wherever possible in the views, such as the `ST_X()`s and `ST_Y()`s, since they are so cheap, and then fully leverage the use of DB triggers to maintain fields that are too expensive to compute on an _ad hoc_ basis. It'll be a column by column call on each, perhaps with use of `EXPLAIN ANALYZE` to put some numbers to the cost of the computations. Again, I think this would be work in that [aforementioned issue here](https://github.com/cityofaustin/atd-vz-data/pull/1288#discussion_r1337797768).

Reply:
Cool, yes, after talking about this in the standup meeting, I totally agree that keep this view slim is the way to go. 💯